lib/ecto/repo/schema.ex

```diff
@@ -81,44 +82,33 @@ defmodule Ecto.Repo.Schema do
        when is_list(rows) do
     mapper = init_mapper(schema, dumper, adapter, placeholder_map)
 
-    {rows, {header, has_query?, placeholder_dump, _}} =
-      Enum.map_reduce(rows, {%{}, false, %{}, 1}, fn fields, acc ->
-        {fields, {header, has_query?, placeholder_dump, counter}} = Enum.map_reduce(fields, acc, mapper)
+    {rows, {header, placeholder_dump, _}} =
+      Enum.map_reduce(rows, {%{}, %{}, 1}, fn fields, acc ->
+        {fields, {header, placeholder_dump, counter}} = Enum.map_reduce(fields, acc, mapper)
         {fields, header} = autogenerate_id(autogen_id, fields, header, adapter)
-        {fields, {header, has_query?, placeholder_dump, counter}}
+        {fields, {header, placeholder_dump, counter}}
       end)
 
     header = Map.keys(header)
 
     placeholder_size = map_size(placeholder_dump)
 
-    placeholder_vals_list =
+    {placeholder_cast_params, placeholder_dump_params} =
       placeholder_dump
-      |> Enum.map(fn {_, {idx, _, value}} -> {idx, value} end)
+      |> Enum.map(fn {_, {idx, _, cast_value, dump_value}} -> {idx, cast_value, dump_value} end)
       |> Enum.sort
-      |> Enum.map(&elem(&1, 1))
+      |> Enum.map(&{elem(&1, 1), elem(&1, 2)})
+      |> Enum.unzip
 
-    if has_query? do
-      {rows, value_param_count} = plan_query_in_rows(rows, header, adapter, placeholder_size)
-      {rows, header, placeholder_vals_list, fn -> placeholder_size + value_param_count end}
-    else
-      counter = fn ->
-        Enum.reduce(
-          rows,
-          placeholder_size,
-          &(Enum.count(&1, fn {_, val} -> not match?({:placeholder, _}, val) end) + &2)
-        )
-      end
-
-      {rows, header, placeholder_vals_list, counter}
-    end
+    {rows, cast_params, counter} = plan_query_in_rows(rows, header, adapter, placeholder_size)
```

Comment:
This is the biggest compromise that had to be made to accommodate this change. When none of the rows contains a query, we still have to traverse and re-order the fields in the rows here.

There is another option that can be done to avoid this. We can send the casted parameters to the adapter without re-ordering them and add the logic for re-ordering them in the adapter. The downside of this is that we are putting more of the cast parameter logic into the adapters for `insert_all` when none of the other query methods are doing so.

Reply:
I assume that before it was responsibility of the adapters to sort the fields? If we are sorting them here now, does it mean it less work to do on the adapters too?

Reply:
Before this PR the fields would be sorted once in Ecto and once in the adapters if any of the rows had a query for a value. And only sorted in the adapters if there weren't any query value.

Ecto needed to sort them in order to have the correct query parameter count when planning the row queries. If there were no queries then it wouldn't be sorted in Ecto. 

The adapters needed to sort them as well because rows could be missing fields present in other rows that need to have their [values converted to `nil`](https://github.com/elixir-ecto/ecto_sql/blob/master/lib/ecto/adapters/sql.ex#L838) so that they can have their. value [replaced by the keyword `DEFAULT`](https://github.com/elixir-ecto/ecto_sql/blob/master/lib/ecto/adapters/postgres/connection.ex#L241). 

I originally wanted to take the sorting out of the adapter but it would mean changing the value of the rows going into the adapter. Which would be ok for anything that uses `Ecto.Adapters.SQL` but not custom ones that aren't using it.

The other option I see is to pass the unsorted cast parameters into the adapter and then once inside the adapter zip them with the rows and sort them here: https://github.com/elixir-ecto/ecto_sql/blob/master/lib/ecto/adapters/sql.ex#L826.

Reply:
Hrm... so what you are saying is that we always keep the params and cast_params in the same order? So if the adapter needs to re-order one, then it reorders both?

Yeah, I actually like that more then. Let's go with it. :)

Reply:
Done :).  The code is a bit more complex because of the need to keep track of which row the cast params are in, but it gets rid of the double sorting when there is no row query.  ecto_sql changes: https://github.com/elixir-ecto/ecto_sql/pull/424